remove paired device

vakar8888 opened the pull request.

tonyofbyteball commented:
Thanks but you cannot just remove it. Various things might stop working and users will have no idea why.

The pairing might be used in multisig wallets or shared addresses.  The removal would break them.  Also, chat messages from the removed correspondent would go nowhere.

I think you should check that the pairing to be removed is not used anywhere.  Second, notify the peer that it was removed, then the peer should remove his end of the pairing automatically (no point leaving around a half-deleted pairing).

vakar8888 commented:
Thanks for your answer.
OK. I check the database and delete the corresponding records. After that I inspect the existing code how to notify the peers and implement the removing of the pairing from the other end. This should be done via eventbus, right? Is it necessary to wait for a handshake/commitment from the paired device before deleting?

tonyofbyteball commented:
> how to notify the peers and implement the removing of the pairing from the other end. This should be done via eventbus, right?

eventBus is for internal events only.  I think you'll have to use something like 

device.sendMessageToDevice(device_address, "text", message, ..)

but use another (new) message type instead of "text".  You'll receive and handle this message on the other end in handleMessageFromHub https://github.com/byteball/byteballcore/blob/master/wallet.js#L129

> Is it necessary to wait for a handshake/commitment from the paired device before deleting?
No, the paired device might never reply.

vakar8888 commented:
Thanks for your answer.
I've analysed your code and the database to find out how devices are used.
I'm going to implement 'remove device' in that way, that only devices not used in multisig wallets can be removed, and the peer will be notified after that.

Are there any difference between the databases (byteball.sqlite and byteball-light.sql)?

Wallets can't be deleted. Right?
If so, it should be possible to hide and unhide wallets (and devices). So the user can choose what he want's to see.

tonyofbyteball commented:
> I've analysed your code and the database to find out how devices are used.
> I'm going to implement 'remove device' in that way, that only devices not used in multisig wallets can be removed, and the peer will be notified after that.

correct

> Are there any difference between the databases (byteball.sqlite and byteball-light.sql)?

The differences are not related to devices.

> Wallets can't be deleted. Right?
> If so, it should be possible to hide and unhide wallets (and devices). So the user can choose what he want's to see.

Deleting a wallet is irreversible, all its coins are lost (if it had any).

vakar8888 commented:
- only devices not used in multisig wallets can be removed
- peer will be notified
- requires commits in byteballcore

Question: after removing a paired device, pairing again with the same address doesn't work. Any idea what is missing?

tonyofbyteball commented on src/js/controllers/correspondentDevices.js:
```diff
@@ -56,26 +58,65 @@ angular.module('copayApp.controllers').controller('correspondentDevicesControlle
-		}, 100);
+
+		var device = require('byteballcore/device.js');
+
```
it makes sense to check that it is removable, just to be safe

tonyofbyteball commented on src/js/controllers/correspondentDevices.js:
```diff
@@ -56,26 +58,65 @@ angular.module('copayApp.controllers').controller('correspondentDevicesControlle
+			ifError: function(error){}}
+			);
+
+		device.removeCorrespondentDevice(addr, function() {
```
to be consistent in naming, addr -> device_address

vakar8888 commented:
> the 3rd argument should be body than can be serialized using JSON.stringify(). Why functions?
fixed; added body with content 'removed'

> using .map you can do it in one line
fixed

> it makes sense to check that it is removable, just to be safe
fixed

> to be consistent in naming, addr -> device_address
fixed

tonyofbyteball commented on src/js/controllers/correspondentDevices.js:
```diff
@@ -56,31 +58,69 @@ angular.module('copayApp.controllers').controller('correspondentDevicesControlle
+
+		// check to be safe
+		correspondentListService.deviceCanBeRemoved(device_address, {
+			ifOk: function(){
```
When a device cannot be removed, this situation doesn't seem like an "error".
Why not just return a boolean, wrapping the callbacks into an object seems unnecessary

vakar8888 commented:
Before I fix this, I'm waiting for your comments concerning 'remove devices' on the pull request of byteballcore.

vakar8888 commented:
message renamed to removed_paired_device; refactoring deviceCanBeRemoved
